<a href="https://colab.research.google.com/github/GabeMaldonado/DS4A_repo/blob/main/Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Team 5
## Data Cleaning II Notebook

In this notebook we will built on what we coded in the Data Cleaning I assignment and enhance the data cleaning process for our data.

Please note that the dataset was moved to a github repository and we will retrieve the data from its URL using pandas.

In [1]:
# import required lobraries
import numpy as np
import pandas as pd



In [2]:
# Configuring data output options
# To read all characters in the column
pd.set_option('display.max_colwidth', None)

# to display all rows
pd.set_option('display.max_rows', None)

In [3]:
# Load up the data dictionary as well to reference the columns and keys to facilitate feature selection

# Load the data dictionary to df called data_dict
data_dict_link = 'https://github.com/GabeMaldonado/DS4A_repo/blob/main/DS4A_Project/Datasets/Loan%20Data/Data_Dictionary.xlsx?raw=True'

data_dict = pd.read_excel(data_dict_link)
data_dict



LoanStatNew                                                                                                                                                                                                         Description
0                    acc_now_delinq                                                                                                                                                     The number of accounts on which the borrower is now delinquent.
1              acc_open_past_24mths                                                                                                                                                                          Number of trades opened in past 24 months.
2                        addr_state                                                                                                                                                          The state provided by the borrower in the loan application
3                          all_util                                                                                                                                                                               Balance to credit limit on all trades
4                        annual_inc                                                                                                                                       The self-reported annual income provided by the borrower during registration.
5                  annual_inc_joint                                                                                                                           The combined self-reported annual income provided by the co-borrowers during registration
6                  application_type                                                                                                                Indicates whether the loan is an individual application or a joint application with two co-borrowers
7                       avg_cur_bal                                                                                                                                                                             Average current balance of all accounts
8                    bc_open_to_buy                                                                                                                                                                           Total open to buy on revolving bankcards.
9                           bc_util                                                                                                                               Ratio of total current balance to high credit/credit limit for all bankcard accounts.
10         chargeoff_within_12_mths                                                                                                                                                                              Number of charge-offs within 12 months
11          collection_recovery_fee                                                                                                                                                                                      post charge off collection fee
12       collections_12_mths_ex_med                                                                                                                                                    Number of collections in 12 months excluding medical collections
13                      delinq_2yrs                                                                                                        The number of 30+ days past-due incidences of delinquency in the borrower's credit file for the past 2 years
14                      delinq_amnt                                                                                                                                  The past-due amount owed for the accounts on which the borrower is now delinquent.
15                             desc                     

In [4]:
# Load up load the dataset from the github repo  to df_loan_data
loan_data_link = 'https://github.com/GabeMaldonado/DS4A_repo/blob/main/DS4A_Project/Datasets/Loan%20Data/loan.xlsx?raw=True'
df_loan_data = pd.read_excel(loan_data_link)


In [5]:
# Sample/display 10 random values to get an overall idead of the data
df_loan_data.sample(10)

id  member_id  ...  total_bc_limit  total_il_high_credit_limit
9802    836427    1046461  ...             NaN                         NaN
4205   1006581    1232994  ...             NaN                         NaN
9658    837231    1047351  ...             NaN                         NaN
38940   296903     296900  ...             NaN                         NaN
11112   820510    1028740  ...             NaN                         NaN
26818   562819     724214  ...             NaN                         NaN
26139   573515     737697  ...             NaN                         NaN
5370    986749    1210693  ...             NaN                         NaN
29630   520090     672327  ...             NaN                         NaN
3296   1020958    1249953  ...             NaN                         NaN

[10 rows x 111 columns]

Now that we have the data loaded into a df, we can inspect its datatype as well as counting the `nans`.

In [6]:
# get info about the dataframe: datatype and number of nans
df_loan_data.info(verbose=True, null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 111 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   id                              39717 non-null  int64         
 1   member_id                       39717 non-null  int64         
 2   loan_amnt                       39717 non-null  int64         
 3   funded_amnt                     39717 non-null  int64         
 4   funded_amnt_inv                 39717 non-null  float64       
 5   term                            39717 non-null  object        
 6   int_rate                        39717 non-null  float64       
 7   installment                     39717 non-null  float64       
 8   grade                           39717 non-null  object        
 9   sub_grade                       39717 non-null  object        
 10  emp_title                       37258 non-null  object        
 11  e

Looks like we have lots of empty columns. Empty columnns are not relevant for our analyssis as they contain no value information. These empty columns can be safely dropped from the df.

In [7]:
# get the all the names for the empty columns and put them in a list
df_loan_stats = pd.DataFrame(df_loan_data.count(), columns=['counts'], index=None)
empty_cols = list(df_loan_stats[df_loan_stats['counts'] ==0].index)

In [8]:
empty_cols

['mths_since_last_major_derog',
 'annual_inc_joint',
 'dti_joint',
 'verification_status_joint',
 'tot_coll_amt',
 'tot_cur_bal',
 'open_acc_6m',
 'open_il_6m',
 'open_il_12m',
 'open_il_24m',
 'mths_since_rcnt_il',
 'total_bal_il',
 'il_util',
 'open_rv_12m',
 'open_rv_24m',
 'max_bal_bc',
 'all_util',
 'total_rev_hi_lim',
 'inq_fi',
 'total_cu_tl',
 'inq_last_12m',
 'acc_open_past_24mths',
 'avg_cur_bal',
 'bc_open_to_buy',
 'bc_util',
 'mo_sin_old_il_acct',
 'mo_sin_old_rev_tl_op',
 'mo_sin_rcnt_rev_tl_op',
 'mo_sin_rcnt_tl',
 'mort_acc',
 'mths_since_recent_bc',
 'mths_since_recent_bc_dlq',
 'mths_since_recent_inq',
 'mths_since_recent_revol_delinq',
 'num_accts_ever_120_pd',
 'num_actv_bc_tl',
 'num_actv_rev_tl',
 'num_bc_sats',
 'num_bc_tl',
 'num_il_tl',
 'num_op_rev_tl',
 'num_rev_accts',
 'num_rev_tl_bal_gt_0',
 'num_sats',
 'num_tl_120dpd_2m',
 'num_tl_30dpd',
 'num_tl_90g_dpd_24m',
 'num_tl_op_past_12m',
 'pct_tl_nvr_dlq',
 'percent_bc_gt_75',
 'tot_hi_cred_lim',
 'total_bal

In [9]:
# lets drop the empty columns from our data
# create a new df called df_loan_data_no_mt_cols
df_loan_stats_no_mt_cols = df_loan_data.drop(columns=empty_cols)
df_loan_stats_no_mt_cols.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 57 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   id                          39717 non-null  int64         
 1   member_id                   39717 non-null  int64         
 2   loan_amnt                   39717 non-null  int64         
 3   funded_amnt                 39717 non-null  int64         
 4   funded_amnt_inv             39717 non-null  float64       
 5   term                        39717 non-null  object        
 6   int_rate                    39717 non-null  float64       
 7   installment                 39717 non-null  float64       
 8   grade                       39717 non-null  object        
 9   sub_grade                   39717 non-null  object        
 10  emp_title                   37258 non-null  object        
 11  emp_length                  38642 non-null  object    

After dropping the empty columns, we went from 111 columns to 52. We have significanlty reduced the number of columns by dropping columns that were irrelevant. 
We can see that there are missing values in the remaining columns. We can handle the missing values by imputting the mean of the values for each missing data point. This is the preferred approach at this point. However, we might take a different approach during the exploratory data analysis (EDA) stage of the project. 


In [10]:
df_clean = df_loan_stats_no_mt_cols.fillna(df_loan_stats_no_mt_cols.mean()) #took the average and plugged in values for the NaN's

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: DataFrame.mean and DataFrame.median with numeric_only=None will include datetime64 and datetime64tz columns in a future version.
  """Entry point for launching an IPython kernel.


In [12]:
df_clean.info() #= df.dropna() #dropped rows with NaN's

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 57 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   id                          39717 non-null  int64         
 1   member_id                   39717 non-null  int64         
 2   loan_amnt                   39717 non-null  int64         
 3   funded_amnt                 39717 non-null  int64         
 4   funded_amnt_inv             39717 non-null  float64       
 5   term                        39717 non-null  object        
 6   int_rate                    39717 non-null  float64       
 7   installment                 39717 non-null  float64       
 8   grade                       39717 non-null  object        
 9   sub_grade                   39717 non-null  object        
 10  emp_title                   37258 non-null  object        
 11  emp_length                  38642 non-null  object    

There are a few columns that still contain missing values. Upon closer inspection we can see that such columns are of type `datetime64`. We will look at these variables closer during the EDA stage to determine if the contribute any insights to our analysis. 